flair/models/biomedical_entity_linking.py

```diff
@@ -0,0 +1,988 @@
+                query_embeds=mention_sparse_embeds, dict_embeds=self.dict_sparse_embeds
+            )
+            sparse_ids, sparse_distances = self.retrieve_sparse_topk_candidates(
+                score_matrix=sparse_score_matrix, topk=topk + 10
+            )
+
+            # combine dense and sparse scores
+            hybrid_ids = []
+            hybrid_scores = []
+            # for every embedded mention
+            for (
+                top_dense_ids,
+                top_dense_scores,
+                top_sparse_ids,
+                top_sparse_distances,
+            ) in zip(dense_ids, dense_scores, sparse_ids, sparse_distances):
+                ids = top_dense_ids
+                distances = top_dense_scores
+                for sparse_id, sparse_distance in zip(
+                    top_sparse_ids, top_sparse_distances
+                ):
+                    if sparse_id not in ids:
+                        ids = np.append(ids, sparse_id)
+                        distances = np.append(
+                            distances, sparse_weight * sparse_distance
+                        )
+                    else:
+                        index = np.where(ids == sparse_id)[0][0]
+                        distances[index] = (
+                            sparse_weight * sparse_distance
+                        ) + distances[index]
+
+                sorted_indizes = np.argsort(-distances)
+                ids = ids[sorted_indizes][:topk]
+                distances = distances[sorted_indizes][:topk]
+                hybrid_ids.append(ids.tolist())
+                hybrid_scores.append(distances.tolist())
+
+            return [
+                np.append(self.dictionary[ind], score)
+                for ind, score in zip(hybrid_ids, hybrid_scores)
+            ]
+        # use only dense embeds
+        else:
+            dense_distances, dense_ids = self.dense_dictionary_index.search(
+                x=mention_dense_embeds, k=topk
+            )
+            return [
+                np.append(self.dictionary[ind], score)
+                for ind, score in zip(dense_ids, dense_distances)
+            ]
+
+    def embed_dictionary(
+        self,
+        model_name_or_path: str,
+        dictionary_name_or_path: str,
+        mean_centering: bool,
+        batch_size: int,
+    ):
+        # check for embedded dictionary in cache
+        dictionary_name = os.path.splitext(os.path.basename(dictionary_name_or_path))[0]
+        cache_folder = os.path.join(flair.cache_root, "datasets")
+        file_name = f"bio_nen_{model_name_or_path.split('/')[-1]}_{dictionary_name}"
+        cached_dictionary_path = os.path.join(
+            cache_folder,
+            f"{file_name}.pk",
+        )
+
+        # If exists, load the cached dictionary indices
+        if os.path.exists(cached_dictionary_path):
+            self.load_cached_dictionary(cached_dictionary_path)
+
+        # else, load and embed
+        else:
```

Comment:
Here is where we can us `TextProcess` to rm punctuation and what not, w/ a parameter in the fn